src/PopIII/popiii.cpp

```diff
@@ -0,0 +1,479 @@
+				break;
+			case 11:
+				numdens[n - 1] = userData_.primary_species_11;
+				break;
+			case 12:
+				numdens[n - 1] = userData_.primary_species_12;
+				break;
+			case 13:
+				numdens[n - 1] = userData_.primary_species_13;
+				break;
+			case 14:
+				numdens[n - 1] = userData_.primary_species_14;
+				break;
+
+			default:
+				amrex::Abort("Cannot initialize number density for chem specie");
+				break;
+		}
+	}
+
+	amrex::Real const x0 = prob_lo[0] + 0.5 * (prob_hi[0] - prob_lo[0]);
+	amrex::Real const y0 = prob_lo[1] + 0.5 * (prob_hi[1] - prob_lo[1]);
+	amrex::Real const z0 = prob_lo[2] + 0.5 * (prob_hi[2] - prob_lo[2]);
+
+	// cloud parameters
+	const double R_sphere = userData_.R_sphere;
+	const double omega_sphere = userData_.omega_sphere;
+	const double renorm_amp = userData_.rescale_factor;
+	const double numdens_init = userData_.numdens_init;
+	const double core_temp = userData_.temperature;
+
+	auto const &dvx = userData_.dvx.const_table();
+	auto const &dvy = userData_.dvy.const_table();
+	auto const &dvz = userData_.dvz.const_table();
+
+	amrex::ParallelFor(indexRange, [=] AMREX_GPU_DEVICE(int i, int j, int k) {
+		amrex::Real const x = prob_lo[0] + (i + static_cast<amrex::Real>(0.5)) * dx[0];
+		amrex::Real const y = prob_lo[1] + (j + static_cast<amrex::Real>(0.5)) * dx[1];
+		amrex::Real const z = prob_lo[2] + (k + static_cast<amrex::Real>(0.5)) * dx[2];
+		amrex::Real const r = std::sqrt(std::pow(x - x0, 2) + std::pow(y - y0, 2) + std::pow(z - z0, 2));
+		amrex::Real const distxy = std::sqrt(std::pow(x - x0, 2) + std::pow(y - y0, 2));
+
+		eos_t state;
+		amrex::Real rhotot = 0.0;
+
+		for (int n = 0; n < NumSpec; ++n) {
+			state.xn[n] = numdens[n] * numdens_init;
+			rhotot += state.xn[n] * spmasses[n]; // spmasses contains the masses of all species, defined in EOS
+		}
+
+		// normalize -- just in case
+		std::array<Real, NumSpec> mfracs = {-1.0};
+		Real msum = 0.0;
+		for (int n = 0; n < NumSpec; ++n) {
+			mfracs[n] = state.xn[n] * spmasses[n] / rhotot;
+			msum += mfracs[n];
+		}
+
+		for (int n = 0; n < NumSpec; ++n) {
+			mfracs[n] /= msum;
+			// use the normalized mass fractions to obtain the corresponding number densities
+			state.xn[n] = mfracs[n] * rhotot / spmasses[n];
+		}
+
+		// amrex::Print() << "cell " << i << j << k << " " << rhotot << " " << numdens_init << " " << numdens[0] << std::endl;
+
+		amrex::Real const phi = atan2((y - y0), (x - x0));
+
+		double vx = renorm_amp * dvx(i, j, k);
+		double vy = renorm_amp * dvy(i, j, k);
+		double const vz = renorm_amp * dvz(i, j, k);
+
+		if (r <= R_sphere) {
+			state.rho = rhotot;
+			state.T = core_temp;
+			eos(eos_input_rt, state);
+
+			// add rotation to vx and vy
+			vx += (-1.0) * distxy * omega_sphere * std::sin(phi);
+			vy += distxy * omega_sphere * std::cos(phi);
+
+		} else {
+			state.rho = 0.01 * rhotot;
+			state.p = 3.595730e-10; // pressure equilibrium - this is the pressure within the core
```

Comment:
How did you compute this value?

Reply:
Ah, I pre-calculated it during testing and debugging to make life easier but never updated the code to automatically calculate it. Done now.